# Preparación del problema

## Importación de las librerias 

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np

In [2]:
tips_df = sns.load_dataset('tips')
tips_df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [3]:
tips_df.isnull().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [4]:
tips_df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

## Fastidiamos el dataset

In [5]:
np.random.seed(42)

nan_tip = np.random.choice(tips_df.index, 10, replace=False)
tips_df.loc[nan_tip, 'tip'] = np.nan

negative_tip = np.random.choice(tips_df.index.difference(nan_tip), 3, replace=False)
tips_df.loc[negative_tip, 'tip'] = -tips_df.loc[negative_tip, 'tip'].abs()

nan_total_bill = np.random.choice(tips_df.index, 3, replace=False)
tips_df.loc[nan_total_bill, 'total_bill'] = np.nan

In [6]:
tips_df.isnull().sum()

total_bill     3
tip           10
sex            0
smoker         0
day            0
time           0
size           0
dtype: int64

In [7]:
tips_df.describe()

total_bill         tip        size
count  241.000000  234.000000  244.000000
mean    19.688631    2.938675    2.569672
std      8.848604    1.540356    0.951100
min      3.070000   -4.000000    1.000000
25%     13.280000    2.000000    2.000000
50%     17.780000    2.900000    2.000000
75%     24.060000    3.587500    3.000000
max     50.810000   10.000000    6.000000

# Ejercicio 1 (Inicio)

## Ejercicio 1.1 

Realiza con .copy una copia del dataframe tips_df con el que vas a trabajar.

Llamalo df, de manera que tengas el dataframe con las manipulaciones por si necesitas usarlo o realizar la copia de nuevo. 

Imprime por consola los 5 primeros registros (head) y los 5 últimos (tail), imprime las forma del dataset (shape), el indice (index) y las columnas (columns).

Imprime la información del mismo con un (info) y también los estadísticos con un (describe).

Haz por último un describe que incluya todas las columnas con 'all'.

## Ejercicio 1.2

Comprueba que columnas tienen NaN (isna().sum()), a continuación rellena los NaN de la columna tip con la media de la respectiva columna (fillna).

En el caso de los NaN de total_bill elimina las filas correspondientes.

Busca en tips, los valores que no tienen sentido (negativos) y elimina esas filas (o sobreescribe el dataframe solo con las filas validas, tip >= 0)

## Ejercicio 1.3

Comprueba si el dataset tiene filas duplicadas (duplicated)

(Le he añadido 2 filas duplicadas, podría ser un cliente que ha tenido el mismo comportamiento que otro, pero vamos a presuponer que es un error)

Quiero que en el primer caso, me elimines las duplicadas (todas menos la primera aparición, drop_duplicates).

In [ ]:
df.loc[244] = df.loc[242]

df.loc[245] = df.loc[242]

### Ejercicio 1.3.2 

Ahora quiero que me elimines los duplicados, pero conservando unicamente la última aparición (en lugar de la primera, que realizas por defecto, utiliza keep='last').

In [ ]:
df.loc[245] = df.loc[243]

# Ejercicio 2 

## Ejercicio 2.1

Quiero que me muestres los registros con indice 10, 100 y 200 con (loc)

Quiero que ordenes el dataset a partir de la columna total_bill con orden descendente (sort_values)

Una vez ordenada, quiero que me muestres los 8 primeros registros (iloc)

## Ejercicio 2.2

Convierte la columna smoker con un mappeo donde .map('Yes':1, 'No':0)

Realiza un df.info()

Convierte la columna smoker a booleana (astype) y enseñame las 5 primeras filas


## Ejercicio 2.3 

Crea una columna que se llame porcentaje_tip, a partir del siguiente calculo:

$$ porcentaje\_tip = \frac{tip}{tip + total\_bill}*100 $$

Crea una columna para posteriormente poder borrarla, 

por ejemplo: $$df['col\_a\_borrar'] = 1$$

Por último elimina la columna col_a_borrar (drop)

# Ejercicio 3 

## Ejercicio 3.1

Realiza un filtrado condicional simple, donde obtengas un dataframe llamado fumadores y otro llamado no_fumadores

## Ejercicio 3.2 

Realiza un filtrado condicional múltiple, donde consigas un dataframe llamado fumadores_viernes_o_sabados, donde las condiciones sean:

fumadores, correspondientes al día de viernes o sábado. 

Muestra un describe de fumadores_viernes_o_sabados.

## Ejercicio 3.3

Realiza un nuevo dataset con (.query) llamado hombres_propinas_bajas, donde sex sea 'Male' y porcentaje_tip < 10

Muestra los 5 primeros registros por pantalla.

## Ejercicio 3.4

Crea una máscara booleana, que cumpla los siguientes criterios:

mask 

day: Fri, Sat, Sun (Fin de semana)
total_bill > 30 
size > 1

Utiliza la misma para crear un dataframe llamado imp_grupos_consumidores_finde

Muestra 5 registros por pantalla


# Ejercicio 4

## Ejercicio 4.1

Quiero exportar los datasets en la carpeta donde estoy realizando la ejecución con las siguientes condiciones:

df --> csv (.to_csv)

fumadores --> txt (.to_csv)

no_fumadores --> csv (sin header) (.to_csv)

hombres_propinas_bajas --> html (.to_html)

imp_grupos_consumidores_finde --> excel, quiero que la hoja se llame subgrupo (.to_excel, sheet_name)

df --> json (.to_json)

## Ejercicio 4.2

Carga el dataset fumadores.csv (pd.read_csv) y lo llamas df_fumadores


# Ejercicio 5

## Ejercicio 5.1

Importa el dataset df (según el guardado, quizás necesites index_col)

Realiza los siguientes estudios con groupby

1) La media de tip por day 

2) La media de tip para smoker 

3) La media de tip para sex


## Ejercicio 5.2

Realiza un groupby, donde estudiemos por day y por smoker (juntos), la mediana de tip.

Realiza el groupby para índice day junto a smoker y ayudate de (.agg) para estudiar para tip y total_bill (la media, count, min y max)

## Ejercicio 5.3

Quiero los siguientes pivot_table

1º - index=sex | columns=day | values=tip | aggfunc=mean

2º - index=sex | columns=day | values=tip | aggfunc=mean (quiero que tenga margins y quiero que se llame Total con margins_name)

3º - index=day | columns=sex | values=tip  (quiero que tenga margins y quiero que se llame Total con margins_name) y además 
quiero que tenga aggfunc=['mean', 'median', 'max', 'min']